draft-issuance.md

```diff
@@ -0,0 +1,169 @@
+endeavors to optimize its economic framework, accommodating changing circumstances while
+maintaining predictability and stability in rewards distribution.
+
+This proposal outlines a solution to address challenges associated with the existing block
+subsidy issuance mechanism in the Zcash network. The primary goal of this proposal is to
+introduce a more predictable and stable issuance of ZEC by smoothing out the issuance
+curve while preserving the supply cap. It's important to note that this proposal does
+not seek to alter the fundamental aspects of Zcash's issuance policy. The average block
+subsidy size over time will remain the same and the funds for block subsidies will last
+a similar amount of time. Instead, it focuses solely on enhancing the predictability
+and consistency of the block subsidy issuance process.
+
+Smoothing the emissions curve helps ensure that the network remains economically
+viable and stable as it transitions from a traditional issuance mechanism to one
+that maintains a sustainable and predictable issuance of rewards over time. It
+prevents abrupt changes in the rate of newly issued coins, which could lead to
+disruptions in the network's economic model and potentially impact its security
+and sustainability. A smoother emissions curve allows for a more gradual and controlled
+transition, providing ZEC stakeholders and participants with a clear understanding of
+how rewards will be distributed over time.
+
+
+
+# Specification
+
+Smoothing the issuance curve is possible using an exponential decay formula that
+satisfies the following requirements:
+
+## Requirements
+
+1. Block subsidies MUST be weakly decreasing
+2. Block subsidies SHOULD approximate a continuous function
+3. When `AVAILABLE_SUBSIDIES(h) > 0` then block subsidies for block `h`
+MUST always be `> 0`, preventing a final “unmined” zatoshi
+4. For any 4 year period, all paid out block subsidies MUST equal approximately
+half of `AVAILABLE_SUBSIDIES` at the beginning of that 4 year period
+5. This functionality MUST be introduced as part of a network upgrade
+
+The above requirements assume no deflationary action, i.e. that no ZEC is added
+to `AVAILABLE_SUBSIDIES`. They are referenced below as **Rn**.
+
+## Solution
+
+Given the block height `h` define a function **S**, such that:
+
+**S(h)** = Block subsidy for a given `h`, that satisfies above requirements.
+
+Please note that
+
+`AVAILABLE_SUBSIDIES(h+1) = AVAILABLE_SUBSIDIES(h) - S(h)` assuming no deflationary action.
+
+An exponential decay function **S** satisfies **R1** and **R2** above:
+
+`S(h) = BLOCK_SUBSIDY_FRACTION * AVAILABLE_SUBSIDIES(h)`
+
+Finally, to satisfy **R3** above we need to always round up to at least 1 Zatoshi
+if `AVAILABLE_SUBSIDIES(h) > 0`:
+
+`S(h) = ROUND_UP(BLOCK_SUBSIDY_FRACTION * AVAILABLE_SUBSIDIES(h))`
+
+# Rationale
+
+## `BLOCK_SUBSIDY_FRACTION`
+
+That value of `41 / 100_000_000` was selected so that it satisfies the equation:
```

Comment:
It might be worth mentioning that `41 / 100_000_000` is 3.1% more than the target issuance rate:
```sh
$ python3 -c 'print((1 - 41/100000000.)**(420768*4))'
0.5015461452386029
```

Reply:
Sorry I had the wrong calculation for the rate difference, it is actually 0.31% less than the target.

If you'd like to make that difference even smaller, some possible fractions and their variance percentages are:
```
$ python3
>>> (1 - ((1 - 410/1000000000.)**(420768*4)) - 0.5)/0.5 * 100
-0.3092290477205717
>>> (1 - ((1 - 411/1000000000.)**(420768*4)) - 0.5)/0.5 * 100
-0.1405433047117155
>>> (1 - ((1 - 412/1000000000.)**(420768*4)) - 0.5)/0.5 * 100
0.02785874818078682
>>> (1 - ((1 - 4118/10000000000.)**(420768*4)) - 0.5)/0.5 * 100
-0.005798995334638235
>>> (1 - ((1 - 4119/10000000000.)**(420768*4)) - 0.5)/0.5 * 100
0.011031292633090217
```

Feel free to keep it as it is, or choose one of these fractions instead.

(We can't go any higher than a 4-digit multiplier, because it could overflow a 63 bit integer if all funds are deposited back into the ZSF. It's unlikely, but we have to think about these things!)

Reply:
These calculations are outdated, they need to be updated to use `PostBlossomHalvingInterval`. This might change some of the constants with larger divisors. (The current constant in the spec is still the closest for that divisor, but it's around 0.45% lower issuance.)

Reply:
@aphelionz @nathan-at-least & jason: do any of the authors want the issuance to be closer to the current issuance? -0.436% is not that much of a difference, but we can make it 100x smaller if you want.

Here is the issuance difference of -0.436% for the current `BLOCK_SUBSIDY_FRACTION`, and two alternative fractions that would give -0.015% and +0.002% issuance differences.

```python
>>> (1 - ((1 - 41/100000000.)**1680000) - 0.5)/0.5 * 100
-0.435650152033773
>>> (1 - ((1 - 4125/10000000000.)**1680000) - 0.5)/0.5 * 100
-0.014704836099199703
>>> (1 - ((1 - 4126/10000000000.)**1680000) - 0.5)/0.5 * 100
0.002096231305492857
```

(Feel free to tag Jason in the comments, I couldn't find his GitHub handle.)

Reply:
`4126/10000000000.` is an interesting one. If that's the absolute closest perhaps we could use that one. I'm flexible tbh.

Reply:
`4126/10000000000` is the closest power-of-10 fraction that achieves the requirements, and is technically reasonable.

Almost all programming languages have 64 bit data types, and some codebases use signed types for amounts, so the number of bits available for this calculation is 63. (Some languages and platforms have 128-bit types, but some don't.)

Here is a rationale for limiting the numerator that can go in the ZIP:

The largest possible amount in the ZSF is `MAX_MONEY`, in the theoretically possible case that all issued funds are deposited back into the ZSF. If this happened, the largest interim sum in the block subsidy calculation would be `MAX_MONEY * 4126 + 10000000000`.

This uses `62.91` bits, which is just under the 63 bit limit for signed 64-bit integer amount types.

The numerator could be brought closer to the limit by using a larger denominator, but the difference in the amount issued would be very small. So we chose a power-of-10 denominator for simplicity.

Reply:
Personal opinion: deltas that small are fine because the whole success of the ZSF is predicated on integrating a large enough volume of deposits that the balance remains steady (or grows) over a given time period. So if the actual issuance rate is a tiny delta from current issuance rate, that would be completely outweighed by deposits.

If, OTOH, deposits are so small that this tiny rate difference matters, then the network is failing to find long term sustainability, so IMO the delta doesn't matter in that case, because the Big Picture is a much worse problem.

Reply:
This is a useful goal and rationale for a small delta over the long term. If you decide to accept a small delta, please put them in the ZIP. However, there could be short-term impacts: if you think those impacts are small, please say that as well.

Here is a potential alternative goal and rationale for a minimal delta over the short term. It doesn't conflict with your rationale over the long term, because it is focused on the short-term impact, rather than the long-term impact.

Goal section:

We want to decrease the short-term impact of the deployment of this ZIP on block reward recipients, and minimise the potential reputational risk to Zcash of changing the block reward amount.

Rationale section:

The chosen `BLOCK_SUBSIDY_FRACTION` changes the block rewards by `+0.002%` over the 4 years after the ZIP activates, which is equivalent to (TODO: 0.XX ZEC per day). This is the smallest change in block rewards we can achieve within technical limitations.

Reply:
Updated the fraction to `4126 / 100_000_000` @tomekpiotrowski please take note

Reply:
Just a note for anyone reading this, the fraction is `4126/10_000_000_000`.

Reply:
This rationale needs to be included in the ZIP, because it justifies that the ZIP is implementable:

> Here is a rationale for limiting the numerator that can go in the ZIP:
> 
> The largest possible amount in the ZSF is `MAX_MONEY`, in the theoretically possible case that all issued funds are deposited back into the ZSF. If this happened, the largest interim sum in the block subsidy calculation would be `MAX_MONEY * 4126 + 10000000000`.
> 
> This uses `62.91` bits, which is just under the 63 bit limit for signed 64-bit integer amount types.
> 
> The numerator could be brought closer to the limit by using a larger denominator, but the difference in the amount issued would be very small. So we chose a power-of-10 denominator for simplicity.

Reply:
I've added it :+1: